walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,119 @@
+        ArrayList<TimeRange> optionalsSlots = new ArrayList<>();
+        if(optionalsEventTimeRangeList.size() > 0) {       
+            optionalsSlots = findSlots(optionalsEventTimeRangeList, request);
+        }
+
+        // Store available TimeRange slots for required attendees.
+        ArrayList<TimeRange> mandatorySlots = new ArrayList<>();
+        if(eventsTimeRangeList.size() > 0) {
+            mandatorySlots = findSlots(eventsTimeRangeList, request);
+        }
+
+        // Finds slots that are common to both requuired and optional attendees and returns them
+        // if there are any.
+        ArrayList<TimeRange> combinedSlots = new ArrayList<>();
+        if(optionalsSlots.size() > 0) {
+            for(TimeRange optionalSlot : optionalsSlots) {
+                for(TimeRange mandatorySlot : mandatorySlots) {
+                    if(mandatorySlot.equals(optionalSlot) || optionalSlot.contains(mandatorySlot)) {
+                        combinedSlots.add(mandatorySlot);
+                    } else if(mandatorySlot.contains(optionalSlot)) {
+                        combinedSlots.add(optionalSlot);
+                    }
+                }
+            }
+            if(combinedSlots.size() > 0) return combinedSlots;
+        }
+        return (mandatorySlots.size() > 0) ? mandatorySlots : optionalsSlots;
+    }
+    // Returns whole day as option if there are no conflicts.
+    return Arrays.asList(TimeRange.WHOLE_DAY);
+  }
+
+  // Finds available TimeRange slots for a potential meeting given a meeting request.
+  private ArrayList<TimeRange> findSlots(ArrayList<TimeRange> eventsTimeRangeList, MeetingRequest request) {
+
+    ArrayList<TimeRange> timeSlots = new ArrayList<>();
+    Collections.sort(eventsTimeRangeList, TimeRange.ORDER_BY_START);
+
+    // Remove nested events.
+    for(int i = 0; i < eventsTimeRangeList.size()-1; i++) {
+        if(eventsTimeRangeList.get(i).contains(eventsTimeRangeList.get(i+1))) {
+            eventsTimeRangeList.remove(eventsTimeRangeList.get(i+1));
+        }
+    }
+
+    // Add START_OF_DAY slot if the first TimeRange is not START_OF_DAY.
+    if(eventsTimeRangeList.get(0).start() != TimeRange.START_OF_DAY) {
+        timeSlots.add( TimeRange.fromStartEnd(TimeRange.START_OF_DAY, eventsTimeRangeList.get(0)
+            .start(), false) );
+    }
+
+    for(int i = 0; i < eventsTimeRangeList.size()-1; i++) {
+        int startTime = eventsTimeRangeList.get(i).end();
+        int endTime   = eventsTimeRangeList.get(i+1).start();
+        long timeBetweenEvents = (long) (endTime - startTime);
+
+        // Checks if there is time overlaps, and if there is enough time to last the given meeting duration.
+        if( (startTime < endTime) && (timeBetweenEvents >= request.getDuration()) ) {
+            timeSlots.add( TimeRange.fromStartEnd(startTime, endTime, false) ); 
+        }        
+    }
+
+    // Add last TimeRange if the last TimeRange is not END_OF_DAY inclusive.
```

Comment:
If you have following ranges:
[8:00, END_OF_DAY)
[9:00, 11:00)
Will it return extra slot of [11:00, END_OF_DAY)?

Reply:
If the ranges you provided are timeranges where the attendees are busy elsewhere and are therefore not available to attend the meeting, it will return [START_OF_DAY, 8:00) only.

[11:00, END_OF_DAY) is not returned as an extra slot since not both of the attendees are available to attend at that time.